source/extensions/filters/common/rbac/matchers.cc

```diff
@@ -0,0 +1,161 @@
+    return std::make_shared<const IPMatcher>(principal.source_ip(), false);
+  case envoy::config::rbac::v2alpha::Principal::IdentifierCase::kHeader:
+    return std::make_shared<const HeaderMatcher>(principal.header());
+  case envoy::config::rbac::v2alpha::Principal::IdentifierCase::kAny:
+    return std::make_shared<const AlwaysMatcher>();
+  default:
+    NOT_REACHED;
+  }
+}
+
+AndMatcher::AndMatcher(const envoy::config::rbac::v2alpha::Permission_Set& set) {
+  for (const auto& rule : set.rules()) {
+    matchers_.push_back(Matcher::create(rule));
+  }
+}
+
+AndMatcher::AndMatcher(const envoy::config::rbac::v2alpha::Principal_Set& set) {
+  for (const auto& id : set.ids()) {
+    matchers_.push_back(Matcher::create(id));
+  }
+}
+
+bool AndMatcher::matches(const Network::Connection& connection) const {
+  for (const auto& matcher : matchers_) {
+    if (!matcher->matches(connection)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool AndMatcher::matches(const Network::Connection& connection,
+                         const Envoy::Http::HeaderMap& headers) const {
+  for (const auto& matcher : matchers_) {
+    if (!matcher->matches(connection, headers)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+OrMatcher::OrMatcher(
+    const Protobuf::RepeatedPtrField<::envoy::config::rbac::v2alpha::Permission>& permissions) {
+  for (const auto& permission : permissions) {
+    matchers_.push_back(Matcher::create(permission));
+  }
+}
+
+OrMatcher::OrMatcher(
+    const Protobuf::RepeatedPtrField<::envoy::config::rbac::v2alpha::Principal>& principals) {
+  for (const auto& principal : principals) {
+    matchers_.push_back(Matcher::create(principal));
+  }
+}
+
+bool OrMatcher::matches(const Network::Connection& connection) const {
+  for (const auto& matcher : matchers_) {
+    if (matcher->matches(connection)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool OrMatcher::matches(const Network::Connection& connection,
+                        const Envoy::Http::HeaderMap& headers) const {
+  for (const auto& matcher : matchers_) {
+    if (matcher->matches(connection, headers)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool HeaderMatcher::matches(const Network::Connection&,
+                            const Envoy::Http::HeaderMap& headers) const {
+  return Envoy::Http::HeaderUtility::matchHeaders(headers, header_);
+}
+
+bool IPMatcher::matches(const Network::Connection& connection) const {
+  const Envoy::Network::Address::InstanceConstSharedPtr& ip =
+      destination_ ? connection.localAddress() : connection.remoteAddress();
+
+  return range_.isInRange(*ip.get());
+}
+
+bool PortMatcher::matches(const Network::Connection& connection) const {
+  const Envoy::Network::Address::Ip* ip = connection.localAddress().get()->ip();
+  return ip && ip->port() == port_;
+}
+
+bool AuthenticatedMatcher::matches(const Network::Connection& connection) const {
+  const auto* ssl = connection.ssl();
```

Comment:
How would this support other authentication method, like JWT-based authentication?
@diemtvu

Reply:
currently it does not but we can extend this in a later PR to other mechanisms. I am already iffy on this existing behavior (which I ported from the ext authz filter), and it can most certainly be improved.